### PR TITLE
[CARBONDATA-3256] Fix that MV datamap doesn't take effect using avg expression and count expression

### PR DESCRIPTION
**Pull request by qiuchenjian**
[Problem] MV datamap doesn't affect in this test case:
```
create table test_table(name string, age int, height int,weight int) stored by 'carbondata'
create datamap test_table_mv using 'mv' as select sum(height),count(age),avg(age),name from test_table group by name
explain select avg(age),name from test_table group by name
```
QueryPlan:
```
*HashAggregate(keys=name#33, functions=count(age#34), avg(cast(age#34 as bigint)))
+- Exchange hashpartitioning(name#33, 200)
+- *HashAggregate(keys=name#33, functions=partial_count(age#34), partial_avg(cast(age#34 as bigint)))
+- *FileScan carbondata default.test_tablename#33,age#34]      
``` 

[Solution] Fix this bug
QueryPlan:
```
*HashAggregate(keys=[name#282], functions=[sum(count(age)#280L), sum((avg(age)#281 * cast(count(age)#280L as double))), sum(count(age)#280L)])
+- Exchange hashpartitioning(name#282, 200)
   +- *HashAggregate(keys=[name#282], functions=[partial_sum(count(age)#280L), partial_sum((avg(age)#281 * cast(count(age)#280L as double))), partial_sum(count(age)#280L)])
      +- *Project [count_age#168L AS count(age)#280L, avg_age#169 AS avg(age)#281, test_table_name#170 AS name#282]
         +- *FileScan carbondata default.test_table_mv_table[count_age#168L,avg_age#169,test_table_name#170]]
```

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done

 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10593/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2557/

**Comment by qiuchenjian** [comment]
@jackylk @brijoobopanna @QiangCai @ravipesala  please check and review , thanks very much

**Comment by qiuchenjian** [comment]
@jackylk @brijoobopanna @QiangCai @ravipesala please check and review , thanks very much

**Comment by xuchuanyin** [comment]
affect? Do you mean it does not take effect?

besides, the name of the testfile 'MVQueryPlanTestCase.scala' is too broad. Why not use an existing testfile?

**Comment by qiuchenjian** [comment]
@xuchuanyin i have changed the PR name
becuase almost mv test cases are in MVCreateTestCase, other unmerged PR has changed it,i don't want to get a conflict， so i create a new file

i will open another PR to arrange the MV test cases

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVQueryPlanTestCase.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.test.util.QueryTest
+
+class MVQueryPlanTestCase extends QueryTest {
+
+  test("test avg mv") {
+    sql("drop table if exists test_table")
```
Comment (akashrn5) [inline datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVQueryPlanTestCase.scala]: please can you delete the empty lines after each line?? you can follow same in all the PRs which are open

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11119/

File: datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVQueryPlanTestCase.scala (hunk 1 above)
Comment (qiuchenjian) [reply datamap/mv/core/src/test/scala/org/apache/carbondata/mv/rewrite/MVQueryPlanTestCase.scala]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2859/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3090/

**Comment by kevinjmh** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3013/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11277/

**Comment by ravipesala** [comment]
@qiuchenjian Basically when you provide `sum` and `count` we no need provide `avg` during the creation of datamap. So the following test case should work.
```
sql("create datamap test_table_mv using 'mv' as select sum(height),count(age),name from test_table group by name")
```
On the above datamap following query should hitthat datamap.
```
sql("select avg(age),name from test_table group by name")
```
The actual intention of the code was to derive `avg` from `sum` and `count` columns. So we no need to provide `avg` during datamap creation. Please add the above testcase fix it.

**Comment by qiuchenjian** [comment]
@ravipesala
hi,  
1，illustrate that sum expression is height column, and count avg expression is age column  My mv of test case
2，does you mean that if we use avg expression, we must create mv datamap with sum and count expression ?  But blow testcase is not right in common carbon version

```
  test("test avg mv") {
    sql("drop table if exists test_table")
    sql("create table test_table(name string, age int, height int,weight int) stored by 'carbondata'")
    sql("insert into test_table select 'tom',20,175,130")
    sql("insert into test_table select 'tom',22,185,150")
    sql("create datamap test_table_mv using 'mv' as select sum(age),count(age),name from test_table group by name")
    sql("rebuild datamap test_table_mv")
    val frame = sql("select count(age),avg(age),name from test_table group by name")
    val analyzed = frame.queryExecution.analyzed
    assert(verifyMVDataMap(analyzed, "test_table_mv"))
    checkAnswer(sql("select avg(age),name from test_table group by name"),
      Seq(Row(21.0, "tom")))
    sql("drop table if exists test_table")
  }
```

**Comment by ravipesala** [comment]
Usually while creating mv for avg aggregate we can split it to sum and count aggregate columns. Currently creating MV does not handle the splitting of avg column to sum and count but we should add that support. So if user creates the mn datamap as below 
```
create datamap test_table_mv using 'mv' as select sum(age),count(age),name from test_table group by name
```
or 
```
"create datamap test_table_mv using 'mv' as select avg(age),name from test_table group by name
```

In both above cases we should support below sql
```
select avg(age),name from test_table group by name
```

**Comment by kevinjmh** [comment]
As for ` splitting of avg column to sum and count`,  a optional way could be block avg and tell user to create sum and count instead, so user know what is stored exactly

**Pull request closed without merging**